### ISSUE TITLE
sympify should pass evaluate=False to SymPy functions

### ISSUE THREAD

**Issue by normalhuman** [sympy/sympy#13999]
The flag evaluate=False of `sympify` is meant to prevent evaluation: 
```
>>> sympify("2.2 + 3.3", evaluate=False)
2.2 + 3.3
```
But it doesn't help with functions of a floating point argument:
```
>>> sympify("log(5.5)", evaluate=False)
1.70474809223843
```
This is despite `log`, and other SymPy functions, supporting `evaluate=False` on their own: 
```
>>> log(5.5, evaluate=False)
log(5.5)
```

If ` evaluate=False` is set in `sympify`, it should propagate to any SymPy functions found in the parsed string.

**Comment by jashan498** [comment]
I would like to try this issue : )

**Comment by RISHAV98** [comment]
This problem is happening only with log function ... other function like cos , sin  are not facing this problem..

**Comment by RISHAV98** [comment]
from sympy import cos
>>> sympify(cos(30),evaluate = False)
cos(30)

**Comment by normalhuman** [comment]
```
>>> sympify("cos(30.1)", evaluate=False)
0.252119407926627
```

It happens with floating point arguments for any function.

**Comment by asmeurer** [comment]
This would require every function to support the flag.

**Comment by normalhuman** [comment]
Those that inherit from Function support it. And (in a different issue) it seemed reasonable for sqrt to support it as well.

**Comment by asmeurer** [comment]
It might be better to do it more directly with something like `Basic.__new__(cos, 30.1)`. I'm not fullly sure what the implications of that are, though.

**Comment by jashan498** [comment]
@asmeurer @normalhuman after going through the code, I don't think its quite possible to pass `evaluate=False` flag to SymPy functions(even those inherited from `Function` class). The reason being:
```
>>> type(log(4.6))
<class 'sympy.core.numbers.Float'>
>>> type(cos(30.1))
<class 'sympy.core.numbers.Float'>
```
So whenever we do `sympify("log(5.5)", evaluate=False)`, `log(5.5)` gets passed to `sympify` as `1.70474809223843`, thus leaving no signature of `log` (so there is no `Function` class, hence no possibility of passing  `evaluate=False`). This is not case with integral arguments because `type(log(5)) = log`.

P.S: I think for non-repeating or small floats, if we really need non-evaluated function we can do something like (though not so appealing) :
```
>>> sympify(cos(S(301)/10))
cos(301/10)

```

**Comment by ylemkimon** [comment]
@jashan498 `type(“log(5.5)”) == str`

**Comment by jashan498** [comment]
OH, `evaluate=False` of `sympify` never gets passed to the inner function, not even in case of Integers. Its just `log (Integer (5 )) = 5` and `log (Float ('5.5' ))=1.704`. 
So I suppose there is a need for a method similar to `visit_BinOp` of `EvaluateFalseTransformer` for these function calls too?

#11441 is similar issue I think.

**Comment by asmeurer** [comment]
> So I suppose there is a need for a method similar to visit_BinOp of EvaluateFalseTransformer for these function calls too?

Yes, but the question is how to do it. We can't assume that every function supports evaluate=False. In fact, we don't even know if `f` in `f(x)` is a class or a function. I *think* we can jump directly to `Basic.__new__(f, x)` when `issubclass(f, Basic)`, but there could be issues there (but hey, there are already all kinds of issues with evaluate=False expressions). 

I think we should just try this, and see what happens.

**Comment by asmeurer** [comment]
So one issue is that `Basic.__new__` doesn't call sympify (I never understood the reason for this). So it should actually be `Basic.__new__(f, _sympify(x))`. 

The other concern is that some class will still try to apply something to the input (other than sympifying) when passed `evaluate=False`. For instance, `AssocOp` (`Add` and `Mul`) [remove identity elements](https://github.com/sympy/sympy/blob/eea43ed1983d224b4977abc08fe9e378487411dc/sympy/core/operations.py#L32). I didn't check other classes in the codebase (an audit here would be nice). 

Although maybe this is just an instance of the more general issue with `evaluate=False`, which is that it removes any possible invariants you might be able to say about a given class, as it completely bypasses the constructor. This is why we have issues with evaluate=False expressions all the time, especially in places like the printers, because things assume that some invariant from the constructor is true (for example, that Mul won't have repeated terms). I'm honestly a bit sour on `evaluate=False` in general for this reason, but I think if we are going to do it this way (creating objects bypassing their constructors), we should do it in the simplest and easiest to reason about way possible, which is to skip the subclass constructor completely and only use the Basic constructor (which at least forces it to satisfy the most important invariant of having .args, being hashable, and having sympified args (except not really on the last one). 

I'd love to look more into "wrapping" alternatives to evaluate=False, like what is started at https://github.com/sympy/sympy/blob/eea43ed1983d224b4977abc08fe9e378487411dc/sympy/core/expr.py#L3426, but there are a whole other set of problems to solve to make that work. 

**tl;dr: Let's try adding a `visit_Call` to `EValuateFalseTransformer` that returns `Basic.__new__(f, _sympify(x))` for `f(x)` if `f` is a `Basic` subclass.**

**Comment by jashan498** [comment]
I am struggling to get `x`. I can only get `node` of `x` which is a `Call` object and thus has to be passed through `eval` before using it inside `Basic.__new__`.
Is this the way to get `x` or am I doing it wrong?

**Comment by asmeurer** [comment]
I'm not sure I fully understand your question. However, I've found https://greentreesnakes.readthedocs.io/ to be an invaluable reference when dealing with Python's ast module. Perhaps you can find your answer there.

**Comment by jashan498** [comment]
@asmeurer i am stuck here, the following `diff` although gives `sympify('log(5.5)', evaluate=False) = log(5.5)` but it is unable to pass `evaluate=False` inside `log` (like log(4-3) becomes log(1))
```diff
diff --git a/sympy/parsing/sympy_parser.py b/sympy/parsing/sympy_parser.py
index 4468c8c..2c859f4 100644
--- a/sympy/parsing/sympy_parser.py
+++ b/sympy/parsing/sympy_parser.py
@@ -983,6 +983,20 @@ def flatten(self, args, func):
                 result.append(arg)
         return result
 
+    def visit_Call(self, node):
+
+        if issubclass(sympify(node.func.id),  Basic):
+            args = ast.Call(
+                    func=ast.Name(id='sympify', ctx=ast.Load()),
+                    args=node.args,
+                    keywords=[ast.keyword(arg='evaluate', value=ast.Name(id='False', ctx=ast.Load()))],
+                    starargs=None,
+                    kwargs=None
+                )
+
+            node = ast.Call(func=ast.Attribute(value=ast.Name(id='Basic', ctx=ast.Load()), 
+                        attr='__new__', ctx=ast.Load()),  args=[ast.Name(id=node.func.id, 
+                        ctx=ast.Load()), args],  keywords=[])
+
+        return node

```

**Comment by asmeurer** [comment]
In that case the subtraction needs to be unevaluated, which is not `Call`. But I thought subtractions should already be unevaluated with the current code.

**Comment by asmeurer** [comment]
Why are you calling sympify inside the `Call` visitor? This should not be necessary.

**Comment by jashan498** [comment]
It is a temporary workaround. `node.func.id` is a string and I don't know what's analogous function to check if class represented by the string is  a subclass of 'Basic'

**Comment by jashan498** [comment]
Also, the above code converts `sympify('5', evaluate=False)` to `Basic.__new__(Integer,5)` which gives AttributeError.
ast.dump for the above case:
```
Expression(body=Call(func=Attribute(value=Name(id='Basic', ctx=Load()), 
    attr='__new__', ctx=Load()), args=[Name(id='Integer', ctx=Load()), 
    Call(func=Name(id='sympify', ctx=Load()), args=[Num(n=5)], 
    keywords=[keyword(arg='evaluate', 
    value=Name(id='False', ctx=Load()))])], keywords=[]))
```
Should we use  `FunctionClass` instead of `Basic`?

**Comment by asmeurer** [comment]
You will need to check if the object that will be referenced by the name is a Basic subclass.

**Comment by asmeurer** [comment]
Looks like the code will need to be refactored a bit to pass in the globals dictionary to the transformer, so that it can do a name lookup from there.

**Comment by jashan498** [comment]
@asmeurer  the problem with using `Baisc` contructor is that the format, `Basic.__new__(f, _sympify(x))` ,  is not consistent for every Basic subclass member. 
for example, the following contructors calls(there would be more) give error:
`Basic.__new__(Integer, 2)`
`Basic.__new__(Float, 3)`
`Basic.__new__(Pow, 2, 3)`
`Basic.__new__(Mul ,2, 3)`

**Comment by asmeurer** [comment]
It's `Basic.__new__(cls, *sympify(_args))`. Like `Basic.__new__(Mul, 2, 3)`. 

For atomic types like Integer and Float, however, they don't quite work that way. The arguments aren't stored in `.args`, so that format won't work.

**Comment by jashan498** [comment]
@asmeurer  `Basic.__new__(cls, *sympify(_args))` doesnt work for `Pow`. 
For example `Basic.__new__(Pow, S(3), S(4))` gives error.

**Comment by jashan498** [comment]
Btw have a look at this diff
```diff
diff --git a/sympy/parsing/sympy_parser.py b/sympy/parsing/sympy_parser.py
diff --git a/sympy/parsing/sympy_parser.py b/sympy/parsing/sympy_parser.py
index 4468c8c..1320a71 100644
--- a/sympy/parsing/sympy_parser.py
+++ b/sympy/parsing/sympy_parser.py

@@ -983,7 +986,37 @@ def flatten(self, args, func):
                 result.append(arg)
         return result
 
+    def visit_Call(self, node):
+        class_node = sympify(node.func.id)
+        if issubclass(class_node, Basic) and not issubclass(class_node, AtomicExpr):
+            args = []
+            for m in node.args:
+                if isinstance(m, ast.BinOp):                  # To stop evaluation of arguments like log(4-3)
+                    args.append(self.visit_BinOp(m))
+                elif isinstance(m, ast.Call):                    # To stop evaluation inner ast.Call like log(log(5.5))
+                    args.append(self.visit_Call(m))
+                else:
+                    args.append(ast.Call(
+                    func=ast.Name(id='sympify', ctx=ast.Load()),
+                    args= [m],
+                    keywords=[ast.keyword(arg='evaluate', value=ast.Name(id='False', ctx=ast.Load()))],
+                    starargs=None,
+                    kwargs=None
+                ))
+
+            node = ast.Call(func=ast.Attribute(value=ast.Name(id='Basic', ctx=ast.Load()), 
+                        attr='__new__', ctx=ast.Load()), args=[ast.Name(id=node.func.id, 
+                        ctx=ast.Load()), *args], keywords=[])
+ 
+        return node

```
It gets the job done most of the time (except for Pow and miscellaneous.py functions). Should i open a PR for further discussion?

**Comment by asmeurer** [comment]
> For example Basic.__new__(Pow, S(3), S(4)) gives error.

It does work, but the printing fails, because the assumptions are not initialized correctly. 

```pytb
>>> a = Basic.__new__(Pow, S(3), S(4))
>>> a
Traceback (most recent call last):
  File "./sympy/core/basic.py", line 397, in __repr__
    return sstr(self, order=None)
  File "./sympy/printing/str.py", line 844, in sstr
    s = p.doprint(expr)
  File "./sympy/printing/printer.py", line 236, in doprint
    return self._str(self._print(expr))
  File "./sympy/printing/printer.py", line 274, in _print
    return getattr(self, printmethod)(expr, *args, **kwargs)
  File "./sympy/printing/str.py", line 535, in _print_Pow
    if expr.is_commutative:
AttributeError: is_commutative
```

This is the sort of thing I am talking about with bypassing the constructor.

**Comment by knoahlr** [comment]
Hey, i ran into this same issue. Where **sympify('sin(x)', evaluate=False)** evaluates if x is a float.
I intend to sympify() a string expression before using latex() to produce latex text.
something like.
```python
>>>  expr = 'sqrt(0**2*sin(2.7)**2 + 4*0**2*sin(2.7)**2*cos(2.7)**2 + 0**2*(3*exp(3*2.7)*sin(2.7)*asin(2.7)**3 + 3*exp(3*2.7)*sin(2.7)*asin(2.7)**2/sqrt(-2.7**2 + 1))**2)'
>>>  latex(expr)
'\\sqrt{0^{2} \\cdot 0.42737988023383^{2} + 4 \\left(-0.904072142017061\\right)^{2} \\cdot 0^{2} \\cdot 0.42737988023383^{2} + 0^{2} \\left(0.42737988023383 \\cdot 3 \\left(1.5707963267949 - 1.65019345497948 i\\right)^{3} e^{2.7 \\cdot 3} + \\frac{0.42737988023383 \\cdot 3 \\left(1.5707963267949 - 1.65019345497948 i\\right)^{2} e^{2.7 \\cdot 3}}{\\sqrt{-7.29 + 1}}\\right)^{2}}'

```

You can see from the output that that it evaluates all the trig functions. 
From this thread, I've learnt that this is caused by the **evaluate flag** not being passed to the inner functions. I could use reg ex to add **,evaluate=False** to functions in the expression string.
But if there`s any other way, please help.

**Comment by asmeurer** [comment]
You could use `parse_expr` with a custom token transformer to add the evaluate=False. The problem, as noted here, is that not all functions support evaluate=False. 

You could also subclass the EvaluateFalseTransformer in sympy.parsing.sympy_parser and add the code suggested here, then perform the transformation on the ast.

**Comment by lcontento** [comment]
Hi! I recently ran into this same issue. I can obtain the desired result by using the `evaluate` context-manager from `sympy.core.parameters` (which it looks like it did not exist back then).
```python
import sympy
from sympy.core.parameters import evaluate
with evaluate(False):
    print(sympy.sympify('log(5.5)'))
```
I guess a solution to this issue would then be using this context-manager (removing completely the need for processing the AST). However, in the docstring this context-manager is marked as experimental, so I am not sure if there are some possible issues that may arise from using it. If someone who has more experience with the inner workings of SymPy than me thinks it should be safe, I can make a pull request.

**Comment by oscarbenjamin** [comment]
The global evaluate context manager is not supported by all sympy functions. Neither is the evaluate=False keyword. It could still be used for the cases where it works though
